src/wallet/scriptpubkeyman.h

```diff
@@ -0,0 +1,477 @@
+    //! The public key
+    CPubKey vchPubKey;
+    //! Whether this keypool entry is in the internal keypool (for change outputs)
+    bool fInternal;
+    //! Whether this key was generated for a keypool before the wallet was upgraded to HD-split
+    bool m_pre_split;
+
+    CKeyPool();
+    CKeyPool(const CPubKey& vchPubKeyIn, bool internalIn);
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        int nVersion = s.GetVersion();
+        if (!(s.GetType() & SER_GETHASH))
+            READWRITE(nVersion);
+        READWRITE(nTime);
+        READWRITE(vchPubKey);
+        if (ser_action.ForRead()) {
+            try {
+                READWRITE(fInternal);
+            }
+            catch (std::ios_base::failure&) {
+                /* flag as external address if we can't read the internal boolean
+                   (this will be the case for any wallet before the HD chain split version) */
+                fInternal = false;
+            }
+            try {
+                READWRITE(m_pre_split);
+            }
+            catch (std::ios_base::failure&) {
+                /* flag as postsplit address if we can't read the m_pre_split boolean
+                   (this will be the case for any wallet that upgrades to HD chain split)*/
+                m_pre_split = false;
+            }
+        }
+        else {
+            READWRITE(fInternal);
+            READWRITE(m_pre_split);
+        }
+    }
+};
+
+/*
+ * A class implementing ScriptPubKeyMan manages some (or all) scriptPubKeys used in a wallet.
+ * It contains the scripts and keys related to the scriptPubKeys it manages.
+ * A ScriptPubKeyMan will be able to give out scriptPubKeys to be used, as well as marking
+ * when a scriptPubKey has been used. It also handles when and how to store a scriptPubKey
+ * and it's related scripts and keys, including encryption.
+ */
+class ScriptPubKeyMan
+{
+private:
+    WalletStorage& storage;
+
+protected:
+    std::shared_ptr<WalletDatabase> m_database;
+
+public:
+    ScriptPubKeyMan(WalletStorage& storage) : storage(storage), m_database(storage.GetDatabase()) {}
+
+    const std::string GetDisplayName() const { return storage.GetDisplayName(); }
+    bool IsWalletFlagSet(uint64_t flag) const { return storage.IsWalletFlagSet(flag); }
+    void SetWalletFlag(uint64_t flag) { return storage.SetWalletFlag(flag); }
+    void UnsetWalletFlagWithDB(WalletBatch& batch, uint64_t flag) { return storage.UnsetWalletFlagWithDB(batch, flag); }
+    bool CanSupportFeature(enum WalletFeature wf) const { return storage.CanSupportFeature(wf); }
+    void SetMinVersion(enum WalletFeature wf) { return storage.SetMinVersion(wf); }
+    const CKeyingMaterial& GetEncryptionKey() const { return storage.GetEncryptionKey(); }
+    bool HasEncryptionKeys() const { return storage.HasEncryptionKeys(); }
+    bool IsLocked() const { return storage.IsLocked(); }
```

Comment:
Not asking for this change here since it could be a pain to backport, but in the future I think it'd be nicer to drop these wrapper functions and for keyman implementations to explicitly call storage methods, so the split between what data is managed by the top level wallet class and what data is managed by individual keyman implementations is clearer.

Reply:
Sure. I used the wrappers so that moved code couldn't need to change.